### PR TITLE
Add features and fixes for testem3 comparison

### PR DESCRIPTION
**Pull request by sethrj**
- Ability to disable Rayleigh
- Ability to switch energy deposition axis and change grid via input
- Fix for energy deposition when the end point is coincident with the edep grid (moving back half a step because energy would be placed in the wrong bin at the end of a geometry-limited step)
- Optimization for energy deposition (we're atomic-adding zeroes to a bunch of never-alive tracks)

### CODE REVIEW

File: app/demo-loop/diagnostic/EnergyDiagnostic.hh [hunk 1]
```diff
@@ -118,33 +126,39 @@ void bin_energy(const celeritas::StateHostRef& states, PointersHost& pointers);
...
-    HostItems z_bounds_host;
-    make_builder(&z_bounds_host).insert_back(z_bounds.cbegin(), z_bounds.cend());
-    z_bounds_ = z_bounds_host;
+    HostItems bounds_host;
+    make_builder(&bounds_host).insert_back(bounds.cbegin(), bounds.cend());
+    bounds_ = bounds_host;
 
     // Resize bin data
-    resize(&energy_by_z_, z_bounds_.size() - 1);
+    resize(&energy_per_bin_, bounds_.size() - 1);
```
Comment (amandalund) [inline app/demo-loop/diagnostic/EnergyDiagnostic.hh:147]: Should we add a `fill(real_type(0), &energy_per_bin_)` to be safe? (I noticed I didn't initialize the particle-process diagnostic counts to zero either)

**Review by amandalund — APPROVED** [review]
Looks good, thanks @sethrj!

File: src/sim/detail/InitTracksLauncher.hh [hunk 2]
```diff
@@ -108,6 +112,12 @@ CELER_FUNCTION void InitTracksLauncher<M>::operator()(ThreadId tid) const
         {
             // Initialize it from the position (more expensive)
             geo = init.geo;
+#if !CELER_DEVICE_COMPILE
+            // TODO: kill particle with 'error' state if this happens
+            CELER_VALIDATE(!geo.is_outside(),
+                           << "track started outside the geometry at "
+                           << init.geo.pos);
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/sim/detail/InitTracksLauncher.hh:119]: Yeah, there was a unit error (mm/cm) in the helms file that led to particles starting outside the geometry and segfaulting 🥸

File: app/demo-loop/diagnostic/EnergyDiagnostic.hh (hunk 1 above)
Comment (sethrj) [reply app/demo-loop/diagnostic/EnergyDiagnostic.hh:147]: Yes indeed, I forgot this was the resize that doesn't fill.

**Comment by sethrj** [comment]
This PR is responsible for a substantial change in the time-per-step behavior. In the initial step, all tracks are in the same z plane but are in vacuum and thus deposit no energy. Before this change, all tracks were all hitting the same memory location with an atomic; but after, none of them perform the atomic add.

Before:
Before

After:
After

**Pull request merged**